Allow add-ons to supply version compatibility information in manifests, and use this information to disable incompatible add-ons

LeonarddeR opened the pull request:
### Link to issue number:
Closes #6275

### Summary of the issue:
Add-ons do not supply version compatibility in their manifests. This means that there is no way for NVDA to detect whether an add-on is supported or not.

### Description of how this pull request fixes the issue:
This pr allows setting the following properties in an add-ons manifest:

* minimumNVDAVersion: The minimum required version of NVDA for an add-on to work properly.
* lastTestedNVDAVersion: The last version of NVDA an add-on has been tested with.

This information is acted upon by NVDA as follows:

1. Add-ons that have a higher minimumNVDAVersion in their manifest will be force disabled when NVDA starts. There is no means for the end user to enable the add-on again, except for changing the manifest. If add-on authors require a new version, they do it for a reason.

2. lastTestedNVDAVersion. Users are warned and strongly discouraged from installing / enabling an addon that has not been tested with a version of NVDA equal to or greater than their version.

When one or more add-ons refuse to load due to their incompatible state, the user will receive a warning when NVDA starts. They are able to decide whether to open the add-ons manager to investigate incompatible add-ons and are advised to remove or update them.

GUI warnings are presented:
- During installation of an addon
- Before any installation task, regardless of how initiated (except silent command line installs):
   - launcher
   - auto update check
   - manual update check
   - after postponing an update, and initiating it from either the NVDA menu, the Exit NVDA dialog, the check for update option, or the auto update check.
- After an update of NVDA, if incompatible add-ons are still present. This may occur if a separate windows profile is responsible for the update of NVDA.

### Testing performed:
Added the new keys to an add-ons manifest. The add-on refused to load when the minimum NVDA version is newer.

### Future considerations
1. Personally, I"m a great advocate of prohibiting untested add-ons on secure screens by default, but I think we should not do it right from the start. I know many people who are using Vocalizer 2.0 on their logon screens, and they will from one day on another be unable to do this, unless they update their add-on. We should really consider this again when auto updating is there, though.
2. Add-ons without the `minimumRequiredVersion` manifest key will be treated as supported by default. Once speech refactor lands or at least when the transition to Python 3 starts, we could start enforcing block add-ons missing this manifest entry. We could also make this dependent on what's in the add-on, whether it only contains appModules or also speech synthesizer drivers.

### Known issues with pull request:

### Change log entry:
* Changes
    + Add-on authors now can enforce a minimum required NVDA version for their add-ons. Unsupported add-ons will not load or install. (#6275)
     + Add-on authors must now specify the last version of NVDA the add-on has been tested against. Users will be warned against installing or enabling untested add-ons.
* Changes for developers
    + Add-on authors are now able to provide NVDA version compatibility information in add-on manifests. (#6275)
        * minimumNVDAVersion: The minimum required version of NVDA for an add-on to work properly.
        * lastTestedNVDAVersion: The last version of NVDA an add-on has been tested with.

Commit:
Squashed in ehollig/I7929

Commit:
Allow setting minimum required NVDA version and last tested NVDA version in an add-on manifest, and use them to block add-ons from loading if desired.

Commit:
Review actions after feedback by @josephsl

leonardder commented:
Forgot to point out that I based this on #7930, which is not yet in master. I belief this should not cause any problems though.

michaelDCurran commented:
In the long-term, is the eventual goal that the disable add-ons not 
tested with this version option be on by default?

leonardder commented:
[**@michaelDCurran**](https://github.com/michaelDCurran) commented on [16 feb. 2018 11:25 CET](https://github.com/nvaccess/nvda/pull/8006#issuecomment-366197795 "2018-02-16T10:25:11Z - Replied by Github Reply Comments"):
> In the long-term, is the eventual goal that the disable add-ons not 
> tested with this version option be on by default?

I think we should be very reluctant to do this, at least not at all before we have auto updating of add-ons. Also, if we enforce this, it might be good to do a compatibility check before updating. This shouldn't be very difficult actually.

derekriemer commented:
make sure to update the devGuide before this merges.

Commit:
Updated developer guide

feerrenrut commented on source/addonHandler.py:
```diff
@@ -35,6 +37,11 @@
+#: Supports year.month versions (e.g. 2018.1) and an additional suffix for bug fix releases (e.g. 2018.1.1)
+#: It also matches against strings like 2018.1dev
+#: @type: RegexObject
+NVDA_VERSION_REGEX = re.compile(r"(\d{4,4})\.(\d)(?:\.(\d))?")
```
Some automated tests for the regex would be handy. Makes it a lot easier for someone who may have to modify it in the future to be sure that they haven't regressed for any of the cases that have already been considered.

feerrenrut reviewed (commented):
I think this will need further review, but some quick thoughts looking over this now:
Has the idea of using build number rather than release version been considered? As far as I am aware, the build number is guaranteed to be linear and with no duplicates. It would be simpler to compare (because it is not multi part) and it offers a much finer granularity for addon developers. I haven't thought this through carefully. How do next / RC / dev / try builds fit into the current scheme?

Commit:
Merge remote-tracking branch 'origin/master' into i6275

Commit:
Initial review actions

Commit:
Merge remote-tracking branch 'origin/master' into i6275

feerrenrut commented:
I forgot about this PR, do re-request review if it seems like this happens.

> add-on authors might want to declare their add-ons to be compatible before the actual release comes out.

Yes, good point. However, this is only really necessary for the last tested NVDA version. If we were strict, we would say that you can not test have tested against NVDA release x until release x is built and the build number is known. Being pragmatic about this, what we gain by using the release version is ease for the addon-developers to prepare for an upcoming release and for them to be able to guarantee that the addon will be available from day one of the new version. What we lose is the granularity, and ease of comparison.

I think we should stick with the release version.

derekriemer commented:
is there a way we could make it be disabled for next/master then? We need a way to make sure devs don't have their addon disabled unnecessarily.

leonardder commented:
[**@derekriemer**](https://github.com/derekriemer) commented on [14 mei 2018 09:38 CEST](https://github.com/nvaccess/nvda/pull/8006#issuecomment-388724755 "2018-05-14T07:38:48Z - Replied by Github Reply Comments"):
> is there a way we could make it be disabled for next/master then? We need a way to make sure devs don't have their addon disabled unnecessarily.

You are able to choose whether or not to ignore the last tested version, so I guess devs would enable that.

feerrenrut commented:
> You are able to choose whether or not to ignore the last tested version, so I guess devs would enable that.

Once a user unchecks this ["Disable add-ons that are not tested for the current version of NVDA"](https://github.com/nvaccess/nvda/pull/8006/files#diff-597828d33ba584f58afddcc26b80a992R646), I expect they would be unlikely to ever go and recheck it. They may only be checking this so that they can use one very particular "must have" addon.

I propose we change this mechanism:
- We always check addon versions, unless the addon + addon version + NVDA version combination has been added to a whitelist or blacklist. This ensures that:
  - The decision is made on an addon by addon basis.
  - The user is forced to reconsider the decision after each change in version of either NVDA or the addon.
- Both a whitelist and a blacklist are required, since we must differentiate from "undecided" addons. The default action should be disable.
- The warning prompt should list the problematic addons, the last nvda version they were last tested against, and allow the user to decide to enable or disable the addons. For this we will likely need #7491
- The warning dialog should use strong language to discourage enabling untested addons, with an emphasis on stability 
degradation.
- It should also be possible to get the same information and have the same control from the addons dialog.
- There should not be any way to bypass the disabling of the addon when the NVDA version is less than the minimum NVDA version declared by the addon.

derekriemer commented:
agree.

leonardder commented:
I agree, though it will require a lot of extra work which I might not be able to do myself at the moment.

Commit:
Merge remote-tracking branch 'origin/master' into i6275

leonardder commented:
[**@leonardder**](https://github.com/leonardder) commented on [4 jun. 2018 08:27 CEST](https://github.com/nvaccess/nvda/pull/8006#issuecomment-394247353 "2018-06-04T06:27:44Z - Replied by Github Reply Comments"):
> I agree

That is to say, I have some concerns.

[**@feerrenrut**](https://github.com/feerrenrut) commented on [4 jun. 2018 07:19 CEST](https://github.com/nvaccess/nvda/pull/8006#issuecomment-394236944 "2018-06-04T05:19:33Z - Replied by Github Reply Comments"):
>   - The user is forced to reconsider the decision after each change in version of either NVDA or the addon.

Hmm, now consider the following scenario:

1. A user is using an older, unmaintained add-on that still works, for example Vocalizer version 2.x
2. The user installs NVDA 2018.3 or the version that introduces this feature
3. They have to explicitly enable the Vocalizer add-on, but on that prompt, the add-on is still disabled and therefore people will have to work with Espeak, which they might not be able to understand well enough. Some Espeak languages are not verry well understandable.

It feels a bit like this is a step too far as long as add-on auto updating is not yet implemented. Cc @josephsl 
> - There should not be any way to bypass the disabling of the addon when the NVDA version is less than the minimum NVDA version declared by the addon.

I think that this is the case already.

feerrenrut commented:
While I dont like the Vocalizer example very much, I accept that there is a problem with the proposed solution in the case of a deaf-blind user who has a braille display addon. To cater to this situation, I suggest that before the install is started, the addon versions are checked for compatibility, and the user is warned / given this option to continue with it enabled. To clarify, this would be the same prompt as previously described.

leonardder commented:
There is currently no way to disable an add-on as soon as it is active. Note that disabling an add-on always requires a restart of NVDA. Of course we could also prompt for this restart if one decides to disable an untested add-on.

feerrenrut commented:
No, rather than disable, we would be deciding before starting the new version whether or not it will be allowed to enable. So either it will not become active, or it will be allowed to become active. This will work for auto-update of NVDA, however I'm not sure yet what we can do for a fresh install from a manually downloaded file.

leonardder commented:
I'm sorry, I didn't read your message correctly saying that the prompt would be shown before installation of the new version.

[**@feerrenrut**](https://github.com/feerrenrut) commented on [7 jun. 2018 09:11 CEST](https://github.com/nvaccess/nvda/pull/8006#issuecomment-395317020 "2018-06-07T07:11:18Z - Replied by Github Reply Comments"):
> I'm not sure yet what we can do for a fresh install from a manually downloaded file.

Either a fresh install has no add-ons, or it has been removed previously and the config folder in appdata might still be present, in which case it will be used and we can check add-on compatibility.

Commit:
Merge remote-tracking branch 'origin/master' into i6275, fixing merge conflict

Commit:
Fix dev guide

Commit:
Implement add-on blacklisting and whitelisting, currently only in the addonHandler. Gui work is not yet complete

Commit:
GUI improvements,: use GUIHelper. not quite ready

Commit:
Merge remote-tracking branch 'origin/master' into i6275

Commit:
Move the version info retrieval code from appModuleHandler to fileUtils

Commit:
Add proper file version info to every launcher

Commit:
Add AddonSelectionDialog, far from complete

leonardder commented:
@feerrenrut and @michaeldcurran: I think that there are at least three methods to deal with blacklisting/whitelisting incompatible add-ons pre-installatin:

1. Allow one to blacklist/whitelist before downloading the new version. This will require the update server to suplly the file version of the launcher.
2. On the install/postpone screen, display a warning and change the install button into a button that brings you to a selection dialog that allows you to black/whitelist.
3. When install update is clicked, either from the install/postpone dialog or the startup reminder dialog, display the black/whitelist dialog.

I myself prefer option 2.

Commit:
More work on incompatible add-ons dialog

leonardder commented:
@feerrenrut and @michaeldcurran, is this still on your radar?

If this is planned for 2018.2, there are some nuts to crack regarding UX.

feerrenrut commented:
Are you are happy for me to take a look at the UI for this @leonardder?

leonardder commented:
Yes, please. I think it is work for which a sighted dev is ideal :)

leonardder commented:
[**@feerrenrut**](https://github.com/feerrenrut) commented on [15 aug. 2018 13:20 CEST](https://github.com/nvaccess/nvda/pull/8006#issuecomment-413168635 "2018-08-15T11:20:12Z - Replied by Github Reply Comments"):
> - Within NVDA installer, check addon versions
>   - this may launch the addon-version prompt.

Make sure that this doesn't break silent installations.

> - Before loading an addons when NVDA is starting up, check addon versions
>   - this may launch the addon-version prompt.
>   - This is a fail safe.

Strictly spoken not only a fail safe, but it will also apply to the case where user profile 1 updates NVDA and user profile 2 has incompatible add-ons.

feerrenrut commented:
By the way, I'm going to edit my summary and hide comments that I believe are addressed in it.

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into i6275-addGuiToAddonCompat

leonardder commented:
[**@feerrenrut**](https://github.com/feerrenrut) commented on [16 aug. 2018 05:25 CEST](https://github.com/nvaccess/nvda/pull/8006#issuecomment-413412972 "2018-08-16T03:25:06Z - Replied by Github Reply Comments"):
> > Make sure that this doesn't break silent installations.
> 
> Good point, I assume we should just leave the addons as "undecided" in this case. The user will then get a prompt when they first start up. This isn't exactly optimal for people who might rely on these addons which then may not work.

I agree. Also note that the updater uses the silent installation under the hood.

> > user profile 1 updates NVDA and user profile 2 has incompatible add-ons.
> 
> I hadn't considered multiple windows profiles. I don't think that there is anything we can do about this. Again, thinking about warning users before installation.

They can have the prompt when starting up the new copy for the first time, similar to what happens for silent installations

michaelDCurran commented:
This is looking good. However, two things to consider:
* The NVDA launcher (for manual install/update) recently stopped being able to update the users' configuration. I'm assuming however the add-on version state will be stored in a separate file to the standard NVDA config?
 * If NVDA starts and the dialog is shown to enable some incompatible add-ons, and one of those add-ons is a synthDriver or brailleDisplay that was previously in use, NVDA will obviously fall back to the default synthDriver etc. Once the user has enabled their questionable add-on, what happens then? do they need to restart NVDA? does the config get reloaded? make sure that config is not accidentally saved in this limbo state otherwise once they enable their incompatibile add-on, NVDA will no longer be configured to actually use it.

feerrenrut committed:
moved addonHandler into a package

feerrenrut committed:
Move addon version check to a new file

Adds unit tests for this version check.

feerrenrut committed:
Fix scaling issues in the addon gui

feerrenrut committed:
General code formatting cleanup

feerrenrut committed:
Show unknown compatibility addons dialog on startup

This commit introduces showing the unknown compatibility dialog on
startup if there are addons already installed that are untestsed.

Several other issues are fixed:
- the order of state and file was wrong when saving addon compat state
- AutoWidthColumnCheckListCtrl was stopping NVDA from being able to exit
- Simplified sharing the DPI scaling code

feerrenrut committed:
remove spaces from the end of lines

feerrenrut committed:
Move buttons in add-ons dialog to a vertical list

This makes it easier to add another button.

feerrenrut committed:
Changing compatibility state affects addons dialog

feerrenrut committed:
split the addons dialog actions

The buttons for the selected addons are (vertical) to the right the addons list,
the more general addon actions are (horizontal) below the list

feerrenrut committed:
Fix formatting in about addon dialog

feerrenrut committed:
Prompt before install update

feerrenrut committed:
WIP check addons during installer

feerrenrut committed:
Installer asks about unknown compat addons

feerrenrut committed:
Review actions

for https://github.com/BabbageCom/nvda/pull/9

feerrenrut committed:
Clean up potential accidental wrong NVDA version tuples

feerrenrut committed:
Change wording of whitelisted/blacklisted

feerrenrut committed:
Fix up labels

feerrenrut committed:
Fix up imports

feerrenrut committed:
Use a bitfield for the compatValues.

feerrenrut committed:
Don't show addonCompatGui automatically on secure screens or on AppX

feerrenrut committed:
Make AddonCompatibilityState a singleton

feerrenrut committed:
Ux changes

When installing an addon
- Clarify blocked install message "An updated version of NVDA is
required"
- Require user to acknoledge installation of untested addon
- Reduce verbosity of versions / addon information shown, replace with
addon info button.

Incompatible addon dialog
- Changed wording for clarity

Addon Manager
- Disable "Manage incompatible addons" button when there are no
incompatible addons

feerrenrut committed:
Fix several UX issues

feerrenrut committed:
I6275 add gui to addon compat (#9)

* moved addonHandler into a package

* Move addon version check to a new file

Adds unit tests for this version check.

* Fix scaling issues in the addon gui

* General code formatting cleanup

* Show unknown compatibility addons dialog on startup

This commit introduces showing the unknown compatibility dialog on
startup if there are addons already installed that are untestsed.

Several other issues are fixed:
- the order of state and file was wrong when saving addon compat state
- AutoWidthColumnCheckListCtrl was stopping NVDA from being able to exit
- Simplified sharing the DPI scaling code

* remove spaces from the end of lines

* Move buttons in add-ons dialog to a vertical list

This makes it easier to add another button.

* Changing compatibility state affects addons dialog

* split the addons dialog actions

The buttons for the selected addons are (vertical) to the right the addons list,
the more general addon actions are (horizontal) below the list

* Fix formatting in about addon dialog

* Prompt before install update

* WIP check addons during installer

* Installer asks about unknown compat addons

* Review actions

for https://github.com/BabbageCom/nvda/pull/9

* Clean up potential accidental wrong NVDA version tuples

* Change wording of whitelisted/blacklisted

* Fix up labels

* Fix up imports

* Use a bitfield for the compatValues.

* Don't show addonCompatGui automatically on secure screens or on AppX

* Make AddonCompatibilityState a singleton

* Ux changes

When installing an addon
- Clarify blocked install message "An updated version of NVDA is
required"
- Require user to acknoledge installation of untested addon
- Reduce verbosity of versions / addon information shown, replace with
addon info button.

Incompatible addon dialog
- Changed wording for clarity

Addon Manager
- Disable "Manage incompatible addons" button when there are no
incompatible addons

* Fix several UX issues

Commit:
Merge remote-tracking branch 'origin/master' into i6275, fixing merge conflicts

Commit:
Convert addonHandler/__init__.py to unix style line endings for now to have a way cleaner diff

Commit:
fileUtils: unicode> six.text_type

leonardder commented on source/gui/addonGui.py:
```diff
@@ -5,128 +5,344 @@
+		fd = wx.FileDialog(self, message=_("Choose Add-on Package File"),
 		# Translators: the label for the NVDA add-on package file type in the Choose add-on dialog.
 		wildcard=(_("NVDA Add-on Package (*.{ext})")+"|*.{ext}").format(ext=addonHandler.BUNDLE_EXTENSION),
-		defaultDir="c:",style=wx.FD_OPEN)
```
I wonder whether %userprofile% makes more sense as default dir

feerrenrut replied:
Maybe, but I don't think I'll address that in this PR

leonardder commented on source/gui/addonGui.py:
```diff
@@ -5,128 +5,344 @@
-			bundleName=bundle.manifest['name']
+
+			if not addonVersionCheck.isAddonTested(bundle):
+				if wx.YES != self._showAddonUntestedDialog(bundle):
```
`if self._showAddonUntestedDialog(bundle) == wx.NO:` would be more in line with other code.

leonardder commented on source/gui/addonGui.py:
```diff
@@ -5,128 +5,344 @@
+					addon=bundle,
+					compatibilityStateValue=compatValues.MANUALLY_SET_COMPATIBLE
+				)
+			elif wx.YES != self._showConfirmAddonInstallDialog(bundle):
```
Same as above

leonardder commented on source/gui/addonGui.py:
```diff
@@ -297,29 +548,191 @@ def onEnableDisable(self, evt):
+		except addonHandler.AddonError:
+			log.error("Couldn't change state for %s add-on"%addon.name, exc_info=True)
+			gui.messageBox(
+				# Translators: The message displayed when the state of an add-on cannot be changed.
```
May be we should create separate complete messages for enable and disable, because otherwise some languages might get into grammatical issues.

leonardder commented on source/gui/addonGui.py:
```diff
@@ -297,29 +548,191 @@ def onEnableDisable(self, evt):
+		curAddonsLen=len(self.unknownCompatibilityAddonsList)
+		if curAddonsLen>0:
+			if activeIndex==-1:
+				activeIndex=curAddonsLen-1
```
I do not understand this code. When would an index be -1 and why should we select the last add-on in the lit in this case?

feerrenrut replied:
`activeIndex` is passed in as `-1` from https://github.com/nvaccess/nvda/blob/3d6ad08dfa23d58df50b5660c77b0f2e5af25560/source/gui/addonGui.py#L389 after an addon has been successfully installed. This results in the last addon in the list being selected. I'll add something to the comment for this.

leonardder commented on source/gui/installerGui.py:
```diff
@@ -96,7 +107,25 @@ def __init__(self, parent, isUpdate):
 				# Translators: a message in the installer telling the user NVDA is now located in a different place.
 				msg+=" "+_("The installation path for NVDA has changed. it will now  be installed in {path}").format(path=installer.defaultInstallPath)
-		sHelper.addItem(wx.StaticText(self,label=msg))
+		if shouldAskAboutAddons:
```
I would prefer having this button before continue in the tab order, so that a user will encounter this button before encountering the continue button and may be clicks it accidentally, not knowing that there's another one.

feerrenrut replied:
Yes, I agree. This is actually inconsistent at the moment. I will fix this.

leonardder commented on source/updateCheck.py:
```diff
@@ -359,26 +402,59 @@ def onLaterButton(self, evt):
 		installButton.Bind(wx.EVT_BUTTON, self.onInstallButton)
 		installButton.SetFocus()
-		if storeUpdatesDirWritable:
+		if showAddonCompat:
```
Same concern as earlier, one might miss the button, so it might help to have this before the install button.

leonardder commented on tests/unit/test_addonVersionCheck.py:
```diff
@@ -0,0 +1,344 @@
+#See the file COPYING for more details.
+#Copyright (C) 2018 NV Access Limited
+
+"""Unit tests for the baseObject module, its classes and their derivatives."""
```
Wrong doc string

feerrenrut committed:
Fix bug with "pending disable" in add-on dialog

prevents the status in the add-ons dialog from saying "disabled, pending disable"

feerrenrut committed:
Merge branch 'i6275' of git://github.com/BabbageCom/nvda into i6275-addGuiToAddonCompat

feerrenrut commented on source/addonHandler/addonVersionCheck.py:
```diff
@@ -0,0 +1,186 @@
+			compatValues.MANUALLY_SET_COMPATIBLE,
+			compatValues.MANUALLY_SET_INCOMPATIBLE,
+		]
+		assert compatibilityStateValue in acceptedUserSetCompatVals
```
Yes, I think that's safe, I intended it as a clear way to find the accepted values. I have also updated the docs for the function.

feerrenrut committed:
review actions

leonardder commented:
@feerrenrut: I think we're ready for another review by @michaeldcurran now, right?

michaelDCurran commented:
I wonder if it might be useful to cause the Review add-ons button to be focused by default in all the install / update dialogs where the "However, you have incompatible add-ons" message is displayed. This includes the launcher's install/update dialog where the 'run NvDA on the windows logon screen" checkbox would normally be focused by default. Just another thing to ensure that users are checking this dialog.

michaelDCurran commented on developerGuide.t2t:
```diff
@@ -649,20 +649,26 @@ Generally, it has to be explicitly enabled even in archivers that do support it.
 -
 
+The minimumNVDAVersion and lastTestedNVDAVersion keys are currently not mandatory.
+However, it is highly recommended that these keys be included as they are expected to become mandatory keys in the future.
```
Although the user *can* install an add-on without these keys existing in the manifest, surely from an add-on author's point of view, from now these *are* mandatory?

michaelDCurran commented on source/braille.py:
```diff
@@ -2547,7 +2547,7 @@ def getSerialPorts(filterFunc=None):
 	"""
 	if filterFunc and not callable(filterFunc):
-		raise ValueError("The provided filterFunc is not callable")
+		raise TypeError("The provided filterFunc is not callable")
```
What is the reasoning for this change?

leonardder replied:
addonHandler.getAvailableAddons now accepts a filter function as parameter, similar to here. However, when writing the new code for addonHandler, I discovered that this raised exception deals with a wrong type, not with a wrong value. Therefore it ought to be TypeError there, and I changed it here for consistency reasons.

michaelDCurran reviewed (commented):
The user guide will have to be updated. Especially 3.2 (installing NVDA) and 12.3. Add-ons manager.

Commit:
Dev guide, make new keys mandatory as requested per review

feerrenrut committed:
fix missing translator string

feerrenrut committed:
Fix unit tests

A missing manifest value for minNVDAVersion should cause add-ons
to be considered unsupported.

feerrenrut committed:
Update userguide and developerGuide

feerrenrut commented:
@michaelDCurran 
> This includes the launcher's install/update dialog where the 'run NvDA on the windows logon screen" checkbox would normally be focused by default.

I'm a little uncertain about this. It might make it harder to discover these checkboxes? Would it be better for these dialogs to have a new checkbox similar to the addon installation warning ("I understand that this add-on has not been tested with my version of NVDA") and disable the "install" / "continue" button until it is checked?

michaelDCurran commented:
@feerrenrut  I think a new checkbox would be okay. But the wording would need to be clear. In your example you said "this add-on". In most cases the user is updating NvDA, so it would be more like "Existing add-ons are not compatible with this version of NVDA".

michaelDCurran commented on user_docs/en/userGuide.t2t:
```diff
@@ -1825,18 +1825,33 @@ To disable an add-on, press the "disable" button.
 
+++ Incompatible Add-ons Manager ++[incompatibleAddonsManager]
+The Incompatible Add-ons Manager, which can be accessed via the "manage incompatible add-ons" buttons in the Add-on manager, allows you to override the incompatibility of untested add-ons.
+Add-ons are considered incompatible when they have not been tested with the the version of NVDA that you are using.
```
extra "the" should not be there.

feerrenrut commented on user_docs/en/userGuide.t2t:
```diff
@@ -1825,18 +1825,33 @@ To disable an add-on, press the "disable" button.
+The Incompatible add-ons manager has a short message to explain its purpose as well as the version of NVDA.
+The untested add-ons are presented in a list with the following four columns:
++ Available, a checkbox which reflects whether the add-on can be enabled via the Add-on Manager
++ Package, the name of the add-on
```
I used the plus symbol here to get a numbered list. Since I mentioned there are four columns I thought this might be helpful.

feerrenrut committed:
review actions for userguide

feerrenrut committed:
Add unteseted addon confirmation checkbox

During install the user must check a box to confirm that they understand
that their nvda installation contains addons that may not have been
tested and will be disabled after the installation of NVDA.

feerrenrut committed:
Fixed translation comments test

michaelDCurran commented on source/gui/installerGui.py:
```diff
@@ -96,39 +107,87 @@ def __init__(self, parent, isUpdate):
-		continueButton.SetDefault()
+		if not shouldAskAboutAddons:
+			continueButton.SetDefault()
+			continueButton.SetFocus()
```
I don't think this is right. The focus should start on the Windows logon checkbox, not the continue button.

feerrenrut replied:
Yep, fair point. I was going for consistency with the other dialogs, but I think you are right.

michaelDCurran reviewed (changes requested):
There is something wrong with the line endings of `source/addonHandler/__init__.py` The whole file looks like it has been changed. True it was moved from `source/addonHandler.py` but git diff knows that 75% of the file is the same. I had to read the diff manually with `git diff --ignore-space-at-eol`.

josephsl commented:
Hi, regarding line endings, see my comment on #6275 – this pull request uses Windows line endings as opposed to Unix-style. Thanks.

 

From: Michael Curran <notifications@github.com> 
Sent: Tuesday, December 4, 2018 10:30 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] Allow add-ons to supply version compatibility information in manifests, and use this information to disable incompatible add-ons (#8006)

 

@michaelDCurran requested changes on this pull request.

There is something wrong with the line endings of source/addonHandler/__init__.py The whole file looks like it has been changed. True it was moved from source/addonHandler.py but git diff knows that 75% of the file is the same. I had to read the diff manually with git diff --ignore-space-at-eol.

  _____  

In source/addonHandler/addonVersionCheck.py <https://github.com/nvaccess/nvda/pull/8006#discussion_r238937569> :

> +            # type: (addonHandler.AddonBase, tuple, CompatValues) -> None
+              """Sets the compatibility for an addon. Does not save the value to file.
+              @param NVDAVersion:
+              @param addon: The addon to set compatibility for
+              @param compatibilityStateValue: UNKNOWN allows falling back to auto deduced value. Must be either UNKNOWN,
+              MANUALLY_SET_COMPATIBLE, or MANUALLY_SET_INCOMPATIBLE
+              """
+              cls.initialise()
+              acceptedUserSetCompatVals = [
+                      compatValues.UNKNOWN,
+                      compatValues.MANUALLY_SET_COMPATIBLE,
+                      compatValues.MANUALLY_SET_INCOMPATIBLE,
+              ]
+              assert compatibilityStateValue in acceptedUserSetCompatVals
+
+              autoDeduced = cls._getAutoDeduducedAddonCompat(addon, NVDAVersion)

I'm guessing both the call and the actual definition are a typo: _getAutoDeduducedAddonCompat.
Perhaps _getAutoDeducedAddonCompat

  _____  

In source/gui/installerGui.py <https://github.com/nvaccess/nvda/pull/8006#discussion_r238943331> :

>              # Translators: The label of a button to continue with the operation.
               continueButton = bHelper.addButton(self, label=_("&Continue"), id=wx.ID_OK)
-              continueButton.SetDefault()
+              if not shouldAskAboutAddons:
+                      continueButton.SetDefault()
+                      continueButton.SetFocus()

I don't think this is right. The focus should start on the Windows logon checkbox, not the continue button.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8006#pullrequestreview-181598926> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkNv2i-HkUIhnVhZwTP6aYo4FBM6oks5u12fvgaJpZM4SIALd> .

feerrenrut commented on source/addonHandler/addonVersionCheck.py:
```diff
@@ -0,0 +1,185 @@
+		]
+		assert compatibilityStateValue in acceptedUserSetCompatVals
+
+		autoDeduced = cls._getAutoDeduducedAddonCompat(addon, NVDAVersion)
```
Oh, haha that's a good one. I'll fix this up.

feerrenrut committed:
Focused control changed

When choosing installation options, the focused control should be the
first checkbox, not the continue button.

feerrenrut commented:
@michaelDCurran Indeed, the line endings have changed, however, regardless of what I do with line endings, the only way I can get `git diff` to display a sensible diff is with either `-w` or with `--ignore-space-at-eol`.  Perhaps the diff on GitHub would be better if I reverted the line endings and pushed.

Correct me if I'm wrong, but I believe most of the files are checked in with windows (CRLF) endings, so I expect that changing the line endings is preferable anyway?

> regarding line endings, see my comment on #6275 

@josephsl Sorry, I couldn't find a relevant comment?

feerrenrut commented:
I probably could have been more explicit, the line endings used to be LF (unix style) now they are CRLF (windows style).

michaelDCurran reviewed (approved):
I was able to clearly read the diff for `source/addonHandler/__init__.py` with git diff. 
That file probably should have windows style line endings. But the quick test I did, Git still thought the entire file had changed. It is probably not that important.

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into i6275-addGuiToAddonCompat

feerrenrut committed:
Update changes file for #8006

Add-ons can now supply two new manifest values for better control over
the compatibility of add-ons in NVDA.

leonardder commented:
Woops, I intended to have another run over the code, but turns out I didn't. I also think it is clear that it is pretty difficult to test code that is involved in the update process. Note that this is *not* a request to revert this, but there might be some things to polish.

1. One major disadvantage of the current state is that untested add-ons are disabled by default. While this makes sense, it can result into behavior one might not expect. For example:

    1. Install NVDA 2018.4
    2. Use the Vocalizer (Expressive) add-on along with a desired voice
    3. Whenever 2019.1 is out, perform an update.
    4. As soon as the update process starts, NVDA switches to espeak, even on Windows 10, where a switch to one core might me more desirable.
    5. One might be unable to understand espeak, in which case one is stuck.

    Also consider the impact this might have for braille display add-ons used by deaf blind users.

2. First letter navigation does not work in the incompatible add-ons list. That might be because the check mark is the first item in the list, however I do not recall this behavior from the time I tested the AutoWidthCheckableListCtrl.

leonardder commented:
[**@leonardder**](https://github.com/leonardder) commented on [6 Dec 2018, 13:01 CET](https://github.com/nvaccess/nvda/pull/8006#issuecomment-444848807 "2018-12-06T12:01:47Z - Replied by Github Reply Comments"):
> First letter navigation does not work in the incompatible add-ons list. That might be because the check mark is the first item in the list, however I do not recall this behavior from the time I tested the AutoWidthCheckableListCtrl.

Ugh, and the list control exposes the relevant columns using accDescription. This basically means that disabling object descriptions in NVDA's object presentation settings will only announce the check box, nothing else. Hopefully #8858 can help with this.

michaelDCurran commented:
Switching to eSpeak on Windows 10 needs to be fixed I agree. I was under 
the impression that it would switch to Onecore as that is now the 
default. Clearly there must be another code path.

As for braille displays: There is really nothing that can be done. The 
user is warned pretty strongly that add-ons will be disabled. Perhaps we 
can bluntly state this may affect  speech and braille?

michaelDCurran commented:
This pr is causing alpha snapshots to no longer update. 
After downloading and before installing there is the following traceback:
```
ERROR - unhandled exception (09:21:14.213): 
Traceback (most recent call last): 
  File "gui\__init__.pyc", line 985, in Notify 
  File "updateCheck.pyc", line 578, in _guiExecNotify 
  File "updateCheck.pyc", line 669, in _downloadSuccess 
  File "updateCheck.pyc", line 442, in __init__ 
  File "versionInfo.pyc", line 46, in getNVDAVersionTupleFromString 
ValueError: alpha-16387,0a5c7678 
```
This pr assumes that the version string coming from the update server is numeric, however it is not in the case of snapshots.
I'm not sure if the server actually gives back a numeric version as well or not.
Should we revert this pr for now, based on this and other discussions going on in parallel to improve this feature?

michaelDCurran commented:
Some options:
* We don't bother checking against snapshots. I.e. we just  allow non-numeric versions to accept all add-ons.
* We prepend snapshot version strings with the version year.major like betas and rcs so we can then grab this for checking.
* We some how find out the numeric version from scons or the generated launcher file, and send this to the server when publishing snapshots, and have the numeric version be  given to clients along with the friendly version string. Then the numeric version can be checked by the client.

The first option is the easiest and perhaps should be the way to go for now, but we still could go with one of the other options at any point.

Note: right now I have disabled further alphas from being picked up by update clients.
